### DISCUSSION TITLE
How can I automatically render list items when they go into frame?

### DISCUSSION THREAD

**Discussion by eamonburns** [fyne-io/fyne#4748] [Q&A] [answered]
I have created a list of todo items like so:
```go
todoList := widget.NewList(
    func() int {
        return len(todos)
    },
    func() fyne.CanvasObject {
        return widget.NewCheck("template", func(checked bool) {})
    },
    func(lii widget.ListItemID, co fyne.CanvasObject) {
        check := co.(*widget.Check)

        check.Text = todos[lii].Text
        check.Checked = todos[lii].Completed
    },
)
```

And then I set the content of the window

```go
content := container.NewPadded(todoList)
w.SetContent(content)
```

But, all the todo items are like this

- [ ] template

Until I hover over them with my mouse (or tap them when in the mobile simulation).
How can I automatically render them when they are either in view, or the moment `todoList` is rendered?

**Comment by andydotxyz** [comment, marked as answer]
you are forgetting to call `check.Refresh()` after you manipulate the state.

**Reply by eamonburns** [reply]
Thanks, that worked. Is that something that should be mentioned in [the documentation](https://docs.fyne.io/collection/list)?

So, instead of this (from documentation):
```go
func(i widget.ListItemID, o fyne.CanvasObject) {
    o.(*widget.Label).SetText(data[i])
}
```
It would be this:
```go
func(i widget.ListItemID, o fyne.CanvasObject) {
    o.(*widget.Label).SetText(data[i])
    o.Refresh()
}
```
Or is it left out on purpose?

**Reply by andydotxyz** [reply]
It is left out on purpose because when you call a method it will refresh automatically for you.
When you directly edit state it needs to be told to refresh to get the update pushed to the renderer.
